libusb/os/darwin_usb.c

```diff
@@ -637,6 +637,40 @@ static int darwin_get_config_descriptor(struct libusb_device *dev, uint8_t confi
   return (int) len;
 }
 
+static int darwin_get_platform_device_id (libusb_device *dev, char *data, int length) {
+  struct darwin_cached_device *priv = DARWIN_CACHED_DEVICE(dev);
+  io_iterator_t deviceIterator;
+  io_service_t service;
+  io_string_t path;
+  kern_return_t kresult;
+  size_t s;
+
+  kresult = usb_setup_device_iterator (&deviceIterator, priv->location);
+  if (kresult != kIOReturnSuccess)
+    return darwin_to_libusb (kresult);
+
+  service = IOIteratorNext (deviceIterator);
+  if (service == IO_OBJECT_NULL) {
+    IOObjectRelease(deviceIterator);
+    return LIBUSB_ERROR_NOT_FOUND;
+  }
+
+  kresult = IORegistryEntryGetPath(service, kIOServicePlane, path);
+
+  IOObjectRelease(service);
+  IOObjectRelease(deviceIterator);
+
+  if (kresult != kIOReturnSuccess)
+    return darwin_to_libusb (kresult);
+
+  s = strlen(path) + 1;
+  if (s > length)
+    return LIBUSB_ERROR_OVERFLOW;
+
+  memcpy(data, path, s);
+  return LIBUSB_SUCCESS;
+}
+
 /* check whether the os has configured the device */
```

Comment:
Seems a little overkill to go through the IO registry again. Would be better to just cache the path when the device is detected.

Reply:
BTW: something we've learned in HIDAPI - RegistryPath is not always unique. We've seen different devices with the same path, but different Registry IDs:
https://github.com/libusb/hidapi/commit/4d63a5085ac6e6a5be6f0f29a42a4e739da30881
Not sure is this is applicable to LIBUSB (I believe those where BLE devices), but still, something to consider.